### PR TITLE
Support GHC 9.4

### PR DESCRIPTION
**Pull request by parsonsmatt**
This PR supports GHC 9.4. Unfortunately, we have to vendor some code in for now from GHC, since the relevant functionality is not exposed.

Relevant GHC ticket: https://gitlab.haskell.org/ghc/ghc/-/issues/21599

### CODE REVIEW

**Comment by nomeata** [comment]
Thanks. I wonder if we should just remove that “trying to detect if optimizations are off” thing… all it does is print a warning, right?

File: src/Test/Inspection/Plugin.hs [hunk 1]
```diff
@@ -318,14 +414,27 @@ resultToExpr (Failure s) = App <$> dcExpr 'Failure <*> mkStringExpr s
 proofPass :: UponFailure -> ReportingMode -> ModGuts -> CoreM ModGuts
 proofPass upon_failure report guts = do
     dflags <- getDynFlags
-    let noopt = optLevel dflags < 1
+    let noopt =
+#if MIN_VERSION_ghc(9,4,0)
+            snd $ updOptLevelChanged 1 dflags
```
Comment (phadej) [inline src/Test/Inspection/Plugin.hs:419]: It's not unheard to do `-fno-full-laziness`. Won't this be a `False` then, though enough optimizations are on for `inspection-testing` to work.

I hope this regression is fixed before GHC-9.4 is released, so we won't need to hack around it.

Comment (nomeata) [reply src/Test/Inspection/Plugin.hs:419]: Right, but this discussion shows that trying to guess whether the user intended to run the tests or not is not very helpful in the first place.

**Comment by Bodigrim** [comment]
> I wonder if we should just remove that “trying to detect if optimizations are off” thing…

+1, it's not a plugin's concern. There is no robust way to detect it optimization level: one can enable `-O1`, but disable pretty much every other optimization indivisually.

**Pull request merged**